src/main/java/mate/academy/service/impl/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,53 @@
+package mate.academy.service.impl;
+
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import mate.academy.exception.AuthenticationException;
+import mate.academy.exception.RegistrationException;
+import mate.academy.lib.Inject;
+import mate.academy.lib.Service;
+import mate.academy.model.User;
+import mate.academy.service.AuthenticationService;
+import mate.academy.service.UserService;
+import mate.academy.util.HashUtil;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        if (isNullOrEmpty(email, password)) {
+            throw new AuthenticationException("Email or password shouldn't be null or empty");
+        }
+        Optional<User> user = userService.findByEmail(email);
+        if (authenticate(user, password)) {
+            throw new AuthenticationException("Email or password is invalid");
+        }
+        return userService.findByEmail(email).orElseThrow(() ->
+                new NoSuchElementException("Can't get user by email: " + email));

```

Comment:
why do you need to find one more time? You have a variable `user` that contains information